core/src/storage/collections/btree_map/impls.rs

```diff
@@ -0,0 +1,1963 @@
+    Flush,
+};
+use core::{
+    borrow::Borrow,
+    cmp::Ord,
+};
+#[cfg(feature = "ink-generate-abi")]
+use ink_abi::{
+    HasLayout,
+    LayoutField,
+    LayoutStruct,
+    StorageLayout,
+};
+use scale::{
+    Codec,
+    Decode,
+    Encode,
+};
+#[cfg(feature = "ink-generate-abi")]
+use type_metadata::Metadata;
+
+/// Each node in the tree has 2 * B children.
+///
+/// For the moment we use the same number as the Rust standard library
+/// implementation here. There is no empiric evidence for the effectiveness
+/// of this number yet -- we just chose this number to have a first starting
+/// point.
+pub(super) const B: usize = 6;
+
+/// Number of elements (i.e. key/value pairs) which can be stored in one
+/// node of the tree. The `- 1` is because there needs to be an edge to
+/// the right of the last element.
+///
+/// ```no_compile
+/// const B: usize = 2;
+/// const CAPACITY: usize = 2 * B - 1;
+/// keys  = [    a,    b,    c    ];
+/// edges = [ 1,    2,    3,    4 ];
+/// ```
+pub(super) const CAPACITY: usize = 2 * B - 1;
+
+/// Pointer to the index where a key/value pair is stored in `kv_pairs`.
+pub(super) type KVStoragePointer = u32;
+
+/// The node type, either a `Leaf` (a node without children) or `Internal`
+/// (a node with children). This distinction makes it easier to handle e.g.
+/// cases where nodes are merged or implementing recursion.
+pub(super) enum HandleType {
+    Leaf,
+    Internal,
+}
+
+/// This enum is used when recursively processing an underfull node (i.e. a node has so
+/// few elements that it could be merged with another one). This happens when a key/value
+/// pair is removed from a node.
+///
+/// This enum is returned as a result of `.handle_underfull_node`, which contains
+/// the underfull handling logic. Based on this result the recursion either proceeds or stops.
+enum UnderflowResult {
+    /// We are at the root now.
+    AtRoot,
+    /// The node has an empty parent, referred to by the `NodeHandle` in this tuple.
+    EmptyParent(NodeHandle),
+    /// The elements of an underfull node were merged into another node. The contained
+    /// `NodeHandle` points to a parent node. The right node was merged into the left
+    /// node and subsequently removed.
+    Merged(NodeHandle),
+    /// If merging is not possible we "steal" an element from one node and put it into
+    /// its neighboring node.
+    Stole(NodeHandle),
+}
+
+/// Mapping stored in the contract storage.
+///
+/// This implementation follows the algorithm used by the Rust
+/// `BTreeMap` standard library implementation. The Rust implementation
+/// is in-memory, whereas this implementation uses the ink! storage
+/// primitives (`SyncChunk`, etc.).
+#[cfg_attr(feature = "ink-generate-abi", derive(Metadata))]
+pub struct BTreeMap<K, V> {
+    /// Stores densely packed general BTreeMap information.
+    header: storage::Value<BTreeMapHeader>,
+
+    /// Contains the nodes of the tree.
+    entries: SyncChunk<InternalEntry<K, V>>,
```

Comment:
I renamed `entries` to `nodes`, but your other suggestion doesn't work out since `InternalEntry<K, V>` encapsulates our recycling storage mechanism:
```
enum InternalEntry<K, V> {
    /// A vacant entry pointing to the next vacant index.
    Vacant(Option<NodeHandle>),
    /// An occupied entry contains a tree node with its elements.
    Occupied(Node<K, V>),
}
```
But as you see there is already a `Node<K, V>` struct which encapsulates what you most probably wanted to get at.